datafusion/physical-plan/src/sorts/sort.rs

```
@@ -529,6 +523,12 @@ impl ExternalSorter {
     /// Sorts the in-memory batches and merges them into a single sorted run, then writes
     /// the result to spill files.
     async fn sort_and_spill_in_mem_batches(&mut self) -> Result<()> {
+        if self.in_mem_batches.is_empty() {
```

This was my first version to fix the internal error. Later I decided to move the logic to the caller (`ExternalSorter::insert_batch`) because the name `ExternalSorter::sort_and_spill_in_mem_batches` kind of implies that it only makes sense to call when there are in-mem batches.

Now we handle this case in two places. I can remove the handling here unless a more defensive approach is desired.

---

> Now we handle this case in two places. I can remove the handling here unless a more defensive approach is desired.

I think it's better to be more defensive, I'd prefer to throw an internal error saying that we must assume the buffer is not empty when calling this function.